### PR TITLE
Add Spark extension example

### PR DESCRIPTION
**Pull request by xyang16**
Spark extension poc:

User can call `val outputDf = transformer.transform(df)` to run inference.

### CODE REVIEW

**Comment by xyang16** [comment]
Running result:

[Image_Classification_Spark2.html.zip](https://github.com/deepjavalibrary/djl-demo/files/9912828/Image_Classification_Spark2.html.zip)

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import ai.djl.training.util.ProgressBar
+import org.apache.spark.sql.Row
+
+@SerialVersionUID(123456789L)
+class SparkModel(val url : String) extends Serializable {
+
+  private lazy val criteria = Criteria.builder
+    .setTypes(classOf[Row], classOf[Classifications])
+    .optModelUrls(url)
+    .optTranslator(new SparkImageClassificationTranslator())
```
Comment (lanking520) [inline apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala:27]: is this translator serializable? If not, how do we make sure SparkImageClassificationTranslator being used in each executor?

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import ai.djl.inference.Predictor
+import ai.djl.modality.Classifications
+import ai.djl.repository.zoo.{Criteria, ModelZoo}
+import ai.djl.training.util.ProgressBar
+import org.apache.spark.sql.Row
+
+@SerialVersionUID(123456789L)
+class SparkModel(val url : String) extends Serializable {
+
+  private lazy val criteria = Criteria.builder
```
Comment (lanking520) [inline apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala:24]: given we are passing this by SparkPredictor, this doesn't need to be lazy right?

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    .setTypes(classOf[Row], classOf[Classifications])
+    .optModelUrls(url)
+    .optTranslator(new SparkImageClassificationTranslator())
+    .optProgress(new ProgressBar)
+    .build()
+  private lazy val model = ModelZoo.loadModel(criteria)
```
Comment (lanking520) [inline apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala:30]: same applies here

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkPredictor.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+class SparkPredictor(override val uid: String) extends Transformer {
+
+  def this() = this(Identifiable.randomUID("SparkPredictor"))
+
+  final val inputCol = new Param[String](this, "inputCol", "The input column")
+  final val outputCol = new Param[String](this, "outputCol", "The output column")
+  final val modelUrl = new Param[String](this, "modelUrl", "The model URL")
+
+  def setInputCol(value: String): this.type = set(inputCol, value)
```
Comment (lanking520) [inline apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkPredictor.scala:29]: Why use setter? Can we use builder pattern to create it

Comment (lanking520) [reply apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkPredictor.scala:29]: or Scala can use case class

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/ImageClassificationExample.scala [hunk 5]
```diff
@@ -12,91 +12,28 @@
...
-      val predictor = model.newPredictor()
-      partition.map(row => {
-        // image data stored as HWC format
-        predictor.predict(row).toString
-      })
-    })(Encoders.STRING)
-    println(result.collect().mkString("\n"))
+    println(df.select("image.origin", "image.width", "image.height").show(truncate = false))
+
+    val predictor = new SparkPredictor()
```
Comment (lanking520) [inline apache-spark/spark3.0/image-classification/src/main/scala/com/examples/ImageClassificationExample.scala:32]: how does customer passing their own translator?

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala (hunk 1 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala]: Yes it should be serializable

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/ImageClassificationExample.scala (hunk 5 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/src/main/scala/com/examples/ImageClassificationExample.scala:32]: Changed

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkPredictor.scala (hunk 4 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkPredictor.scala:32]: This is a common pattern for class that extends Transformer, see https://github.com/apache/spark/blob/master/mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala#L162

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala (hunk 2 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala]: OK

File: apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala (hunk 3 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/src/main/scala/com/examples/SparkModel.scala]: Done

File: apache-spark/notebook/Image_Classification_Spark.ipynb [hunk 6]
```diff
@@ -82,38 +89,38 @@
...
-    "\n",
-    "      val height = ImageSchema.getHeight(row)\n",
-    "      val width = ImageSchema.getWidth(row)\n",
-    "      val channel = ImageSchema.getNChannels(row)\n",
-    "      var image = ctx.getNDManager.create(ImageSchema.getData(row), new Shape(height, width, channel)).toType(DataType.UINT8, true)\n",
-    "      // BGR to RGB\n",
-    "      image = image.flip(2)\n",
-    "      pipeline.transform(new NDList(image))\n",
-    "    }\n",
+    "// Translator: a class used to do preprocessing and post processing\n",
```
Comment (frankfliu) [inline apache-spark/notebook/Image_Classification_Spark.ipynb:92]: Should we use spark extension in the jupyter notebook?

File: apache-spark/spark3.0/image-classification/build.gradle [hunk 7]
```diff
@@ -1,32 +1,46 @@
...
 dependencies {
-    implementation platform("ai.djl:bom:0.12.0")
-    implementation "org.apache.spark:spark-sql_2.12:3.0.1"
-    implementation "org.apache.spark:spark-mllib_2.12:3.0.1"
-    implementation "ai.djl:api"
-
-
-    runtimeOnly "ai.djl.pytorch:pytorch-model-zoo"
-    runtimeOnly "ai.djl.pytorch:pytorch-native-auto"
+    implementation "org.apache.spark:spark-core_2.12:${spark_version}"
```
Comment (frankfliu) [inline apache-spark/spark3.0/image-classification/build.gradle:16]: spark extension should have covered spack dependencies

File: apache-spark/spark3.0/image-classification/build.gradle [hunk 8]
```diff
@@ -1,32 +1,46 @@
...
+    implementation "org.apache.spark:spark-sql_2.12:${spark_version}"
+    implementation "org.apache.spark:spark-mllib_2.12:${spark_version}"
+    implementation "ai.djl:api:${djl_version}"
```
Comment (frankfliu) [inline apache-spark/spark3.0/image-classification/build.gradle:19]: We should use bom

File: apache-spark/spark3.0/image-classification/build.gradle [hunk 9]
```diff
@@ -1,32 +1,46 @@
...
+    runtimeOnly "ai.djl.pytorch:pytorch-engine:${djl_version}"
+    runtimeOnly "ai.djl.pytorch:pytorch-native-cpu-precxx11:1.12.1"
 }
 
 compileScala {
     scalaCompileOptions.setAdditionalParameters(["-target:jvm-1.8"])
 }
 
 application {
+    sourceCompatibility = JavaVersion.VERSION_1_8
```
Comment (frankfliu) [inline apache-spark/spark3.0/image-classification/build.gradle:31]: Why not use JDK 11?

File: apache-spark/spark3.0/image-classification/build.gradle [hunk 10]
```diff
@@ -1,32 +1,46 @@
 plugins {
     id 'scala'
     id 'application'
+    id 'com.github.johnrengelman.shadow' version '7.0.0'
```
Comment (frankfliu) [inline apache-spark/spark3.0/image-classification/build.gradle:4]:
```suggestion
    id 'com.github.johnrengelman.shadow' version '7.1.2'
```

File: apache-spark/spark3.0/image-classification/build.sbt [hunk 11]
```diff
@@ -7,6 +7,7 @@ scalacOptions += "-target:jvm-1.8"
 
 resolvers += Resolver.jcenterRepo
 
+libraryDependencies += "org.apache.spark" %% "spark-core" % "3.0.1"
 libraryDependencies += "org.apache.spark" %% "spark-sql" % "3.0.1"
 libraryDependencies += "org.apache.spark" %% "spark-mllib" % "3.0.1"
 libraryDependencies += "ai.djl" % "api" % "0.12.0"
```
Comment (frankfliu) [inline apache-spark/spark3.0/image-classification/build.sbt:13]: Should we upgrade to latest version?

File: gradle.properties [hunk 12]
```diff
@@ -14,3 +14,5 @@ systemProp.org.gradle.internal.publish.checksums.insecure=true
 commons_cli_version=1.5.0
 log4j_slf4j_version=2.18.0
 rapis_version=22.04.0
+spark_version=3.2.2
+djl_version=0.19.0
```
Comment (frankfliu) [inline gradle.properties:18]: this should be 0.20.0-SNAPSHOT

File: apache-spark/notebook/Image_Classification_Spark.ipynb (hunk 6 above)
Comment (xyang16) [reply apache-spark/notebook/Image_Classification_Spark.ipynb:92]: OK

File: apache-spark/spark3.0/image-classification/build.gradle (hunk 7 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/build.gradle]: OK

File: apache-spark/spark3.0/image-classification/build.gradle (hunk 8 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/build.gradle]: OK

File: apache-spark/spark3.0/image-classification/build.gradle (hunk 10 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/build.gradle]: OK

File: apache-spark/spark3.0/image-classification/build.sbt (hunk 11 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/build.sbt]: OK

File: gradle.properties (hunk 12 above)
Comment (xyang16) [reply gradle.properties]: OK

File: apache-spark/spark3.0/image-classification/build.gradle (hunk 9 above)
Comment (xyang16) [reply apache-spark/spark3.0/image-classification/build.gradle:28]: Because the java version on EMR is 8.

**Review by lanking520 — APPROVED** [review]

**Pull request merged**